pkg/middleware/stored_session.go

```diff
@@ -159,11 +182,75 @@ func (s *storedSessionLoader) refreshSession(rw http.ResponseWriter, req *http.R
 	err = s.store.Save(rw, req, session)
 	if err != nil {
 		logger.PrintAuthf(session.Email, req, logger.AuthError, "error saving session: %v", err)
-		return fmt.Errorf("error saving session: %v", err)
+		err = fmt.Errorf("error saving session: %v", err)
+	}
+	return err
+}
+
+func (s *storedSessionLoader) handleObtainLockError(req *http.Request, session *sessionsapi.SessionState) error {
+	wasRefreshed, err := s.checkForConcurrentRefresh(session, req)
+	if err != nil {
+		logger.Errorf("Unable to wait for obtained lock: %v", err)
+		return err
+	}
+	if !wasRefreshed {
+		return errors.New("unable to obtain lock and session was also not refreshed via concurrent request")
+	}
+	return nil
+}
+
+func (s *storedSessionLoader) updateSessionFromStore(req *http.Request, session *sessionsapi.SessionState) error {
+	sessionStored, err := s.store.Load(req)
+	if err != nil {
+		return fmt.Errorf("unable to load updated session from store: %v", err)
+	}
+
+	if sessionStored == nil {
+		return fmt.Errorf("no session available to udpate from store")
 	}
+	*session = *sessionStored
+
 	return nil
 }
 
+func (s *storedSessionLoader) waitForPossibleSessionLock(session *sessionsapi.SessionState, req *http.Request) (bool, error) {
+	var wasLocked bool
+	isLocked, err := session.PeekLock(req.Context())
+	for isLocked {
+		wasLocked = true
+		// delay next peek lock
+		time.Sleep(SessionLockPeekDelay)
+		isLocked, err = session.PeekLock(req.Context())
+	}
+
+	if err != nil {
+		return false, err
+	}
+
+	return wasLocked, nil
+}
+
+// checkForConcurrentRefresh returns true if the session is already refreshed via a concurrent request.
+func (s *storedSessionLoader) checkForConcurrentRefresh(session *sessionsapi.SessionState, req *http.Request) (bool, error) {
+	wasLocked, err := s.waitForPossibleSessionLock(session, req)
```

Comment:
What happens if the lock expires before this function is called? Then we would get wasLocked false, and so wouldn't attempt to reload the session right? I think there's a race still because we aren't making every request to refresh go through the obtain lock dance.

If we loop over every refresh request grabbing the lock, we can make sure that it has the latest state and check whether it needs a refresh under lock. With the current approach, if it doesn't obtain the lock in the first instance, there's a possibility that the refresh fails when we could have retried.

1. A Obtains lock
2. B fails to obtain lock
3. A releases lock
4. B peaks and see session is not locked
5. B checkForConcurrentRefresh returns false, nil
6. B returns an error

Reply:
> If we loop over every refresh request grabbing the lock, we can make sure that it has the latest state and check whether it needs a refresh under lock.

I don't understand what you mean by looping over every refresh request?

Reply:
Do you mean looping over refreshSession method?
This would then be our retry mechanism. How many retries you would do? Or should this be configurable?

Reply:
I updated my code.

I added a retry mechanism for getting validated session and updated the tests to have no side effects in peek method.